packages/commerce-sdk-react/src/hooks/useMutation.ts

```diff
@@ -50,6 +52,11 @@ export const useMutation = <
             const netOptions = mergeOptions(hookConfig.client, options)
             const cacheUpdates = hookConfig.getCacheUpdates(customerId, netOptions, data)
             updateCache(queryClient, cacheUpdates, data)
+        },
+        onError(error: any) {
+            // Typescript does not like having promises inside void functions
+            // so we use void to explicitly tell typescript to ignore it
+            void clearAuthStateOnError(error, auth)
```

Comment:
We've discussed this offline and I think this may introduce race conditions, this isn't hooked up to the `Auth` token management lifecycle since it is a callback function that can be called at any time.

Maybe the correct way is to guard the actual `authenticatedMethod` promise, and if that promise throws error `Customer changed credentials`, re-auth as a guest in **the same promise.** so you never run into race conditions, use the Promise is the key!

Reply:
I think this is quite important to get right, so this is a BLOCKING comment.

Reply:
My suggestion in pseudo code

```js
// useMutation.ts
import {
    useMutation as useReactQueryMutation,
} from '@tanstack/react-query'

export const useMutation = (...args) => {
   const clientMethod = hookConfig.method
   const method = clientMethod.catch(auth.handleCredentialChange)
 
 return useReactQueryMutation(method, ...args)
}


// auth.ts
class Auth {
  handleCredentialChange: async (error) => {
      if (error=== "Customer credential is changed") {
        return await reLogin()
     }
     return
  }

}
```

Reply:
Changes have been applied to use promise.catch instead of the onError callback